### PR TITLE
Bach sample changes

### PR DESCRIPTION
**Pull request by thijs-obj**
Change: Do not insert an additional node when calling `get_sample()`. This ensures that the DataFrame returned by `get_unsampled()` is compatible with the original DF.

To make this possible there is a new requirement to users calling `get_sample()` tho: The dataframe must be in a materialized state.

Other changes:
* Moved some code. Might be easiest to review this code per commit. As all code moving is in 59513106b7547d17d97913fade56695a91c459b2 (=first commit), and the actual changes are in ff4a359cddadca97af591580c529b10471931462 and c18aef8694d271706cc811439af5568748cc99ce
* Add some type annotations

### CODE REVIEW

File: bach/bach/dataframe.py [hunk 1]
```diff
@@ -328,14 +328,34 @@ def dtypes(self) -> Dict[str, str]:
         return {column: data.dtype for column, data in self.data.items()}
 
     @property
-    def group_by(self):
+    def group_by(self) -> Optional['GroupBy']:
```
Comment (thijs-obj) [inline bach/bach/dataframe.py:331]: Just added this cause I saw it was missing. But this cause some small ripples in other places to satisfy mypy (which was right to complain)

File: bach/bach/merge.py [hunk 2]
```diff
@@ -220,26 +220,7 @@ def merge(
...
-    :param how: supported values: {‘left’, ‘right’, ‘outer’, ‘inner’, ‘cross’}
-    :param on: optional, column(s) to join left and right on.
-    :param left_on: optional, column(s) from the left df to join on
-    :param right_on: optional, column(s) from the right df/series to join on
-    :param left_index: If true uses the index of the left df as columns to join on
-    :param right_index: If true uses the index of the right df/series as columns to join on
-    :param suffixes: Tuple of two strings. Will be used to suffix duplicate column names. Must make column
-        names unique
-    :return: A new Dataframe. The original frames are not modified.
+    See :py:meth:`bach.DataFrame.merge` for more information.
```
Comment (thijs-obj) [inline bach/bach/merge.py:223]: Unrelated change: Moved the comments to the DataFrame class. As now the comments there say 'see bach.merge.merge', but that's not included in the generated html it seems: https://objectiv.io/docs/modeling/dataframe/bach.DataFrame.merge

File: bach/bach/sample.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    See :py:meth:`bach.DataFrame.get_sample` for more information.
+    """
+    if sample_percentage is None and filter is None:
+        raise ValueError('Either sample_percentage or filter must be set')
+
+    if not df.is_materialized:
+        raise ValueError("Cannot call get_sample on a non-materialized dataframe. "
+                         "Call materialize() first.")
+
+    original_node = df.base_node
```
Comment (thijs-obj) [inline bach/bach/sample.py:32]: This is basically the whole change

File: bach/bach/merge.py (hunk 2 above)
Comment (jansentom) [reply bach/bach/merge.py:223]: good one

File: bach/bach/sample.py [hunk 4]
```diff
@@ -25,7 +25,11 @@ def get_sample(df: DataFrame,
     if sample_percentage is None and filter is None:
         raise ValueError('Either sample_percentage or filter must be set')
 
-    original_node = df.get_current_node(name='before_sampling')
+    if not df.is_materialized:
```
Comment (jansentom) [inline bach/bach/sample.py:28]: there is a groupby check [below](https://github.com/objectiv/objectiv-analytics/pull/318/commits/ff4a359cddadca97af591580c529b10471931462#R41) that can go because of this.

**Review by jansentom — APPROVED** [review]
small comment, rest looks nice!

File: bach/bach/sample.py (hunk 4 above)
Comment (thijs-obj) [reply bach/bach/sample.py:28]: good point. will clean up

File: bach/bach/series/series.py [hunk 5]
```diff
@@ -862,6 +862,8 @@ def _check_unwrap_groupby(self,
         else:
             if isinstance(wrapped, DataFrame):
                 unwrapped = wrapped.group_by
+                if unwrapped is None:
+                    unwrapped = GroupBy([])
```
Comment (thijs-obj) [inline bach/bach/series/series.py:866]: mypy found this `unwrapper is None` case that was not handled. I guessed from the docstring that this `unwrapped = GroupBy([])` would make sense

File: bach/tests/unit/bach/test_expression.py [hunk 6]
```diff
@@ -144,7 +144,7 @@ def test_is_constant():
     assert not WindowFunctionExpression.construct('agg({}, {}, {})', const1, const2, const3).is_constant
 
 
-def test_is_constant():
+def test_is_single_value():
```
Comment (thijs-obj) [inline bach/tests/unit/bach/test_expression.py:147]: unrelated change that I had stashed somewhere. There is already a function called `test_is_constant` above, so this one overwrote that one

Comment (jansentom) [reply bach/tests/unit/bach/test_expression.py:147]: the name was retarded anyway. Mea culpa

File: bach/bach/dataframe.py [hunk 7]
```diff
@@ -328,14 +328,34 @@ def dtypes(self) -> Dict[str, str]:
...
         return copy(self._group_by)
 
+    @property
+    def is_materialized(self) -> bool:
+        """
+        Return true if this DataFrame is in a materialized state, i.e. all information about the
+        DataFrame's values is encoded in self.base_node.
+
+        A DataFrame that's freshly constructed with :py:meth:`from_table`,
+        :py:meth:`from_model`, or :py:meth:`from_pandas` will be in a materialized state. Operations on such
```
Comment (hendrik-obj) [inline bach/bach/dataframe.py:346]: if a table or model contain json type columns it is not in the materialized state as those get cast as jsonb.

**Comment by hendrik-obj** [comment]
good stuff!

File: bach/bach/dataframe.py (hunk 7 above)
Comment (jansentom) [reply bach/bach/dataframe.py:346]: hmm, in that case we should delegate the is_materialized check partially to the series. Most of them will be able to use Thijs' implementation, but the json one should have special treatment in that case

Comment (thijs-obj) [reply bach/bach/dataframe.py:346]: I added comments here and in the docstring of materialize(). And added this as a bug in our tracking document

Comment (thijs-obj) [reply bach/bach/dataframe.py:346]: I'd rather merge this now tho and fix that in a later branch, so other branches can keep moving.
